app-api/routes/trace.py

```diff
@@ -236,66 +234,80 @@ async def annotate_trace(
         session.commit()
         return annotation_to_json(annotation)
 
-# replace all annotations of a certain source
-@trace.post("/{id}/annotations/update")
+
 async def replace_annotations(
+    session: Session,
+    trace_id: str,
+    user_id: UUID,
+    source: str,
+    annotations: List[Dict],
+) -> Dict[str, int]:
+    """
+    Replaces all annotations of a given source with new ones.
+
+    This is used, e.g. to update the analysis model output, once a new run has completed.
+    """
+
+    if not isinstance(annotations, list):
+        raise ValueError("Annotations must be a list")
+    if not isinstance(source, str):
+        raise ValueError("Source must be a string")
+
+    num_deleted = (
+        session.query(Annotation)
+        .filter(
+            Annotation.trace_id == trace_id,
+            Annotation.extra_metadata.op("->>")("source") == source,
+        )
+        .delete()
+    )
+
+    num_inserted = len(annotations)
+
+    for annotation in annotations:
+        new_annotation = Annotation(
+            trace_id=trace_id,
+            user_id=user_id,
+            address=annotation.get("address"),
+            content=str(annotation.get("content")),
+            extra_metadata=annotation.get("extra_metadata"),
+        )
+        session.add(new_annotation)
+
+    session.commit()
+
+    return {"deleted": num_deleted, "inserted": num_inserted}
+
+
```

Comment:
new function to replace all annotations of a certain source with an updated version (e.g. new analysis results).